### PR TITLE
NAS-102285 / 12 / Add ability to specify CPU topology for VM's

### PR DESCRIPTION
**Pull request by sonicaj**
This PR introduces support for specifying CPU topology for VM's

### CODE REVIEW

File: src/middlewared/middlewared/plugins/vm.py [hunk 1]
```diff
@@ -927,6 +941,11 @@ def random_mac(self):
 
         `devices` is a list of virtualized hardware to add to the newly created Virtual Machine.
         Failure to attach a device destroys the VM and any resources allocated by the VM devices.
+
+        Bhyve allows a maximum of of 16 guest virtual CPU's. By default, bhyve configures every virtual CPU as a
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/vm.py]: Lets not mention bhyve, user doesnt need to know implementation details.

Also there is "of of", duplicated

File: gui/vm/models.py [hunk 2]
```diff
@@ -54,6 +54,14 @@ class VM(Model):
         default=1,
         validators=[MinValueValidator(1), MaxValueValidator(16)],
     )
+    cores = models.IntegerField(
+        verbose_name=_('Cores per socket'),
+        default=1,
+    )
+    threads = models.IntegerField(
+        verbose_name=_('No of threads per core'),
```
Comment (william-gr) [inline gui/vm/models.py]: `Number` prob reads better ; )

File: src/middlewared/middlewared/plugins/vm.py [hunk 3]
```diff
@@ -927,6 +941,11 @@ def random_mac(self):
 
         `devices` is a list of virtualized hardware to add to the newly created Virtual Machine.
         Failure to attach a device destroys the VM and any resources allocated by the VM devices.
+
+        FreeNAS allows a maximum of 16 guest virtual CPU's. By default, it configures every virtual CPU as a
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/vm.py]: We should also avoid mentioning FreeNAS, please ;) . Same will show for TrueNAS.

"Maximum of 16 guest virtual CPUs are allowed" or something

Comment (sonicaj) [reply src/middlewared/middlewared/plugins/vm.py]: You are absolutely correct ( again ) :p Apologies, i have corrected it ; )

File: src/middlewared/middlewared/plugins/vm.py [hunk 4]
```diff
@@ -985,10 +1004,17 @@ def random_mac(self):
...
-        vcpus = data.get('vcpus')
+        vcpus = data['vcpus'] * data['cores'] * data['threads']
         if vcpus:
             flags = await self.middleware.call('vm.flags')
-            if flags['intel_vmx']:
+            if vcpus > 16:
+                verrors.add(
+                    f'{schema_name}.vcpus',
+                    'Maximum 16 vcpus are supported. This value is a product '
+                    f'of "{schema_name}.vcpus", "{schema_name}.cores" and "{schema_name}.threads". Please ensure'
```
Comment (a-stjohn) [inline src/middlewared/middlewared/plugins/vm.py]:
```suggestion
                    f'Please ensure the product of "{schema_name}.vcpus", "{schema_name}.cores" and "{schema_name}.threads"'
```

File: src/middlewared/middlewared/plugins/vm.py [hunk 5]
```diff
@@ -985,10 +1004,17 @@ def random_mac(self):
...
+                    'the product of these values is less then 16.'
```
Comment (a-stjohn) [inline src/middlewared/middlewared/plugins/vm.py]:
```suggestion
                    'is less then 16.'
```

**Pull request merged**